File: IotTemplates/DeviceSystemBridgeTemplate/ProjectTemplates/CustomAdapter/AdapterLib/AdapterDevice.cpp
//
// Module Name:
//
//      AdapterDevice.cpp
//
// Abstract:
//
//      AdapterValue, AdapterProperty, AdapterSignal, and AdapterDevice classes implementation.
//
//      Together with Adapter class, the classes declared in this file implement the IAdapter interface.
//
//
#include "pch.h"
#include "Adapter.h"
#include "AdapterDevice.h"

using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;

using namespace BridgeRT;
using namespace DsbCommon;

namespace AdapterLib
{
    //
    // AdapterValue.
    // Description:
    // The class that implements BridgeRT::IAdapterValue.
    //
    AdapterValue::AdapterValue(
        String^ ObjectName,
        Object^ DefaultData // = nullptr
        )
        : name(ObjectName)
        , data(DefaultData)
    {
    }


    AdapterValue::AdapterValue(const AdapterValue^ Other)
        : name(Other->name)
        , data(Other->data)
    {
    }


    uint32 AdapterValue::Set(IAdapterValue^ Other)
    {
        this->name = Other->Name;
        this->data = Other->Data;

        return ERROR_SUCCESS;
    }


    //
    // AdapterProperty.
    // Description:
    // The class that implements BridgeRT::IAdapterProperty.
    //
    AdapterProperty::AdapterProperty(
        String^ ObjectName,
        String^ IfHint
        )
        : name(ObjectName)
        , interfaceHint(IfHint)
    {
    }


    AdapterProperty::AdapterProperty(const AdapterProperty^ Other)
        : name(Other->name)
        , attributes(Other->attributes)
        , interfaceHint(Other->interfaceHint)
    {
    }


    uint32
    AdapterProperty::Set(IAdapterProperty^ Other)
    {
        this->name = Other->Name;

        IAdapterAttributeVector^ otherAttributes = Other->Attributes;

        if (this->attributes.size() != otherAttributes->Size)
        {
            throw ref new InvalidArgumentException(L"Incompatible Adapter Properties");
        }

        for (uint32 attrInx = 0; attrInx < otherAttributes->Size; ++attrInx)
        {
            AdapterValue^ attr = static_cast<AdapterValue^>(this->attributes[attrInx]->Value);

            attr->Set(otherAttributes->GetAt(attrInx)->Value);
        }

        return ERROR_SUCCESS;
    }

    //
    // AdapterAttribute.
    // Description:
    //  The class that implements BridgeRT::IAdapterAttribute.
    //
    AdapterAttribute::AdapterAttribute(String^ ObjectName, E_ACCESS_TYPE accessType, Object^ DefaultData)
    {
        try
        {
            value = ref new AdapterValue(ObjectName, DefaultData);
            Access = accessType;
        }
        catch (OutOfMemoryException^ ex)
        {
            throw;
        }
    }

    AdapterAttribute::AdapterAttribute(const AdapterAttribute^ other)
        : value(other->value)
        , annotations(other->annotations)
        , covBehavior(other->covBehavior)
        , access(other->access)
    {

    }

    //
    // AdapterMethod.
    // Description:
    // The class that implements BridgeRT::IAdapterMethod.
    //
    AdapterMethod::AdapterMethod(
        String^ ObjectName
        )
        : name(ObjectName)
        , result(HRESULT_FROM_WIN32(ERROR_NOT_READY))
    {
    }


    AdapterMethod::AdapterMethod(const AdapterMethod^ Other)
        : name(Other->name)
        , description(Other->description)
        , inParams(Other->inParams)
        , outParams(Other->outParams)
        , result(Other->result)
    {
    }


    void
    AdapterMethod::InputParams::set(IAdapterValueVector^ Params)
    {
        if (Params->Size != this->inParams.size())
        {
            throw ref new InvalidArgumentException(L"Incompatible method input parameters");
        }

        for (uint32 paramInx = 0; paramInx < Params->Size; ++paramInx)
        {
            dynamic_cast<AdapterValue^>(this->inParams[paramInx])->Set(Params->GetAt(paramInx));
        }
    }


    void
    AdapterMethod::SetResult(HRESULT Hr)
    {
        this->result = Hr;
    }


    //
    // AdapterSignal.
    // Description:
    // The class that implements BridgeRT::IAdapterSignal.
    //
    AdapterSignal::AdapterSignal(
        String^ ObjectName
        )
        : name(ObjectName)
    {
    }


    AdapterSignal::AdapterSignal(const AdapterSignal^ Other)
        : name(Other->name)
        , params(Other->params)
    {
    }


    //
    // AdapterDevice.
    // Description:
    // The class that implements BridgeRT::IAdapterDevice.
    //
    AdapterDevice::AdapterDevice(
        String^ ObjectName
        )
        : name(ObjectName)
    {
    }


    AdapterDevice::AdapterDevice(
        const DEVICE_DESCRIPTOR* DeviceDescPtr
        )
        : name(DeviceDescPtr->Name)
        , vendor(DeviceDescPtr->VendorName)
        , model(DeviceDescPtr->Model)
        , firmwareVersion(DeviceDescPtr->Version)
        , serialNumber(DeviceDescPtr->SerialNumer)
        , description(DeviceDescPtr->Description)
    {
    }


    AdapterDevice::AdapterDevice(const AdapterDevice^ Other)
        : name(Other->name)
        , vendor(Other->vendor)
        , model(Other->model)
        , firmwareVersion(Other->firmwareVersion)
        , serialNumber(Other->serialNumber)
        , description(Other->description)
    {
    }


    void
    AdapterDevice::AddChangeOfValueSignal(
        IAdapterProperty^ Property,
        IAdapterValue^ Attribute)
    {
        try
        {
            AdapterSignal^ covSignal = ref new AdapterSignal(Constants::CHANGE_OF_VALUE_SIGNAL);
            covSignal += ref new AdapterValue(Constants::COV__PROPERTY_HANDLE, Property);
            covSignal += ref new AdapterValue(Constants::COV__ATTRIBUTE_HANDLE, Attribute);

            this->signals.push_back(covSignal);
        }
        catch (OutOfMemoryException^ ex)
        {
            throw;
        }
    }
} // namespace AdapterLib

File: IotTemplates/DeviceSystemBridgeTemplate/ProjectTemplates/ManagedCustomAdapter/AdapterLib/AdapterDevice.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using BridgeRT;

namespace AdapterLib
{
    //
    // AdapterValue.
    // Description:
    // The class that implements IAdapterValue from BridgeRT.
    //
    class AdapterValue : IAdapterValue
    {
        // public properties
        public string Name { get; }
        public object Data { get; set; }

        internal AdapterValue(string ObjectName, object DefaultData)
        {
            this.Name = ObjectName;
            this.Data = DefaultData;
        }

        internal AdapterValue(AdapterValue Other)
        {
            this.Name = Other.Name;
            this.Data = Other.Data;
        }
    }

    //
    // AdapterProperty.
    // Description:
    // The class that implements IAdapterProperty from BridgeRT.
    //
    class AdapterProperty : IAdapterProperty
    {
        // public properties
        public string Name { get; }
        public string InterfaceHint { get; }
        public IList<IAdapterAttribute> Attributes { get; }

        internal AdapterProperty(string ObjectName, string IfHint)
        {
            this.Name = ObjectName;
            this.InterfaceHint = IfHint;

            try
            {
                this.Attributes = new List<IAdapterAttribute>();
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal AdapterProperty(AdapterProperty Other)
        {
            this.Name = Other.Name;
            this.InterfaceHint = Other.InterfaceHint;

            try
            {
                this.Attributes = new List<IAdapterAttribute>(Other.Attributes);
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }
    }

    //
    // AdapterAttribute.
    // Description:
    // The class that implements IAdapterAttribute from BridgeRT.
    //
    class AdapterAttribute : IAdapterAttribute
    {
        // public properties
        public IAdapterValue Value { get; }

        public E_ACCESS_TYPE Access { get; set; }
        public IDictionary<string, string> Annotations { get; }
        public SignalBehavior COVBehavior { get; set; }

        internal AdapterAttribute(string ObjectName, object DefaultData, E_ACCESS_TYPE access = E_ACCESS_TYPE.ACCESS_READ)
        {
            try
            {
                this.Value = new AdapterValue(ObjectName, DefaultData);
                this.Annotations = new Dictionary<string, string>();
                this.Access = access;
                this.COVBehavior = SignalBehavior.Never;
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal AdapterAttribute(AdapterAttribute Other)
        {
            this.Value = Other.Value;
            this.Annotations = Other.Annotations;
            this.Access = Other.Access;
            this.COVBehavior = Other.COVBehavior;
        }
    }

    //
    // AdapterMethod.
    // Description:
    // The class that implements IAdapterMethod from BridgeRT.
    //
    class AdapterMethod : IAdapterMethod
    {
        // public properties
        public string Name { get; }

        public string Description { get; }

        public IList<IAdapterValue> InputParams { get; set; }

        public IList<IAdapterValue> OutputParams { get; }

        public int HResult { get; private set; }

        internal AdapterMethod(
            string ObjectName,
            string Description,
            int ReturnValue)
        {
            this.Name = ObjectName;
            this.Description = Description;
            this.HResult = ReturnValue;

            try
            {
                this.InputParams = new List<IAdapterValue>();
                this.OutputParams = new List<IAdapterValue>();
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal AdapterMethod(AdapterMethod Other)
        {
            this.Name = Other.Name;
            this.Description = Other.Description;
            this.HResult = Other.HResult;

            try
            {
                this.InputParams = new List<IAdapterValue>(Other.InputParams);
                this.OutputParams = new List<IAdapterValue>(Other.OutputParams);
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal void SetResult(int ReturnValue)
        {
            this.HResult = ReturnValue;
        }
    }

    //
    // AdapterSignal.
    // Description:
    // The class that implements IAdapterSignal from BridgeRT.
    //
    class AdapterSignal : IAdapterSignal
    {
        // public properties
        public string Name { get; }

        public IList<IAdapterValue> Params { get; }

        internal AdapterSignal(string ObjectName)
        {
            this.Name = ObjectName;

            try
            {
                this.Params = new List<IAdapterValue>();
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal AdapterSignal(AdapterSignal Other)
        {
            this.Name = Other.Name;

            try
            {
                this.Params = new List<IAdapterValue>(Other.Params);
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }
    }

    //
    // AdapterDevice.
    // Description:
    // The class that implements IAdapterDevice from BridgeRT.
    //
    class AdapterDevice :   IAdapterDevice,
                            IAdapterDeviceLightingService,
                            IAdapterDeviceControlPanel
    {
        // Object Name
        public string Name { get; }

        // Device information
        public string Vendor { get; }

        public string Model { get; }

        public string Version { get; }

        public string FirmwareVersion { get; }

        public string SerialNumber { get; }

        public string Description { get; }

        // Device properties
        public IList<IAdapterProperty> Properties { get; }

        // Device methods
        public IList<IAdapterMethod> Methods { get; }

        // Device signals
        public IList<IAdapterSignal> Signals { get; }

        // Control Panel Handler
        public IControlPanelHandler ControlPanelHandler
        {
            get
            {
                return null;
            }
        }
        
        // Lighting Service Handler
        public ILSFHandler LightingServiceHandler
        {
            get
            {
                return null;
            }
        }

        // Icon
        public IAdapterIcon Icon
        {
            get
            {
                return null;
            }
        }


        internal AdapterDevice(
            string Name,
            string VendorName,
            string Model,
            string Version,
            string SerialNumber,
            string Description)
        {
            this.Name               = Name;
            this.Vendor             = VendorName;
            this.Model              = Model;
            this.Version            = Version;
            this.FirmwareVersion    = Version;
            this.SerialNumber       = SerialNumber;
            this.Description        = Description;

            try
            {
                this.Properties  = new List<IAdapterProperty>();
                this.Methods     = new List<IAdapterMethod>();
                this.Signals     = new List<IAdapterSignal>();
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal AdapterDevice(AdapterDevice Other)
        {
            this.Name               = Other.Name;
            this.Vendor             = Other.Vendor;
            this.Model              = Other.Model;
            this.Version            = Other.Version;
            this.FirmwareVersion    = Other.FirmwareVersion;
            this.SerialNumber       = Other.SerialNumber;
            this.Description        = Other.Description;

            try
            {
                this.Properties = new List<IAdapterProperty>(Other.Properties);
                this.Methods = new List<IAdapterMethod>(Other.Methods);
                this.Signals = new List<IAdapterSignal>(Other.Signals);
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }

        internal void AddChangeOfValueSignal(
            IAdapterProperty Property,
            IAdapterValue Attribute)
        {
            try
            {
                AdapterSignal covSignal = new AdapterSignal(Constants.CHANGE_OF_VALUE_SIGNAL);

                // Property Handle
                AdapterValue propertyHandle = new AdapterValue(
                                                    Constants.COV__PROPERTY_HANDLE,
                                                    Property);

                // Attribute Handle
                AdapterValue attrHandle = new AdapterValue(
                                                    Constants.COV__ATTRIBUTE_HANDLE,
                                                    Attribute);

                covSignal.Params.Add(propertyHandle);
                covSignal.Params.Add(attrHandle);

                this.Signals.Add(covSignal);
            }
            catch (OutOfMemoryException)
            {
                throw;
            }
        }
    }
}
